libs/EXTERNAL/libezgl/src/graphics.cpp

```diff
@@ -702,35 +710,78 @@ void renderer::draw_arc_path(point2d center,
     cairo_stroke(m_cairo);
 }
 
-void renderer::draw_surface(surface *p_surface, point2d top_left)
+void renderer::draw_surface(surface *p_surface, point2d point, double scale_factor)
 {
   // Check if the surface is properly created
   if(cairo_surface_status(p_surface) != CAIRO_STATUS_SUCCESS)
     return;
 
-  // pre-clipping
-  double s_width = (double)cairo_image_surface_get_width(p_surface);
-  double s_height = (double)cairo_image_surface_get_height(p_surface);
+  // calculate surface width and height in screen coordinates
+  double s_width = (double)cairo_image_surface_get_width(p_surface) * scale_factor;
+  double s_height = (double)cairo_image_surface_get_height(p_surface) * scale_factor;
+
+  // calculate surface width and height in world coordinates
+  if (current_coordinate_system == WORLD) {
+    s_width *= m_camera->get_world_scale_factor().x;
+    s_height *= m_camera->get_world_scale_factor().y;
+  }
 
-  if(rectangle_off_screen({{top_left.x, top_left.y - s_height}, s_width, s_height}))
+  // Calculate the top left point
+  point2d top_left = point;
+  if (horiz_justification == justification::center)
+    top_left.x -= s_width/2;
+  else if (horiz_justification == justification::right)
+    top_left.x -= s_width;
+  if (vert_justification == justification::center)
+    top_left.y += (current_coordinate_system == WORLD) ? s_height/2 : -s_height/2;
+  else if (vert_justification == justification::bottom)
+    top_left.y += (current_coordinate_system == WORLD) ? s_height : -s_height;
+
+  if (rectangle_off_screen({{top_left.x, top_left.y - s_height}, s_width, s_height}))
     return;
 
-  // transform the given top_left point
+  // transform the given point
   if(current_coordinate_system == WORLD)
     top_left = m_transform(top_left);
 
+  if (scale_factor != 1) {
+    // save the current state to undo the scaling
+    cairo_save(m_cairo);
+
+    // scale the cairo context with the given scale factor
+    cairo_scale(m_cairo, scale_factor, scale_factor);
+
+    // adjust the corner point based on the context scaling
+    top_left.x /= scale_factor;
+    top_left.y /= scale_factor;
+  }
+
   // Create a source for painting from the surface
   cairo_set_source_surface(m_cairo, p_surface, top_left.x, top_left.y);
 
   // Actual drawing
   cairo_paint(m_cairo);
+
+  if (scale_factor != 1) {
+    // restore the old state to undo the performed scaling
+    cairo_restore(m_cairo);
+  }
 }
 
 surface *renderer::load_png(const char *file_path)
 {
   // Create an image surface from a PNG image
   cairo_surface_t *png_surface = cairo_image_surface_create_from_png(file_path);
 
+  cairo_status_t status = cairo_surface_status(png_surface);
+
+  if (status == CAIRO_STATUS_FILE_NOT_FOUND) {
+    g_warning("load_png: File %s not found.", file_path);
+  }
+  else if (status != CAIRO_STATUS_SUCCESS) {
+    g_warning("load_png: Error loading file %s.", file_path);
```

Comment:
For consistency with the other error message, I think we should change these g_warnings to "renderer::load_png: ..." 
(put renderer at start).
This is my code I think so I'm commenting on myself :).